pkg/cloudprovider/providers/vsphere/vsphere_util.go

```diff
@@ -63,3 +80,112 @@ func GetgovmomiClient(cfg *VSphereConfig) (*govmomi.Client, error) {
 	client, err := newClient(context.TODO(), cfg)
 	return client, err
 }
+
+// Get list of compatible datastores that satisfies the storage policy requirements.
+func (vs *VSphere) GetCompatibleDatastores(ctx context.Context, pbmClient *pbm.Client, resourcePool *object.ResourcePool, storagePolicyID string) ([]mo.Datastore, error) {
+	datastores, err := vs.getAllAccessibleDatastoresForK8sCluster(ctx, resourcePool)
+	if err != nil {
+		return nil, err
+	}
+	var hubs []pbmtypes.PbmPlacementHub
+	for _, ds := range datastores {
+		hubs = append(hubs, pbmtypes.PbmPlacementHub{
+			HubType: ds.Type,
+			HubId:   ds.Value,
+		})
+	}
+	req := []pbmtypes.BasePbmPlacementRequirement{
+		&pbmtypes.PbmPlacementCapabilityProfileRequirement{
+			ProfileId: pbmtypes.PbmProfileId{
+				UniqueId: storagePolicyID,
+			},
+		},
+	}
+	res, err := pbmClient.CheckRequirements(ctx, hubs, nil, req)
+	if err != nil {
+		return nil, err
+	}
+	compatibleHubs := res.CompatibleDatastores()
+	// Return an error if there are no compatible datastores.
+	if len(compatibleHubs) < 1 {
+		return nil, fmt.Errorf("There are no compatible datastores: %+v that satisfy the storage policy: %+q requirements", datastores, storagePolicyID)
+	}
+	var compatibleDatastoreRefs []types.ManagedObjectReference
+	for _, hub := range compatibleHubs {
+		compatibleDatastoreRefs = append(compatibleDatastoreRefs, types.ManagedObjectReference{
+			Type:  hub.HubType,
+			Value: hub.HubId,
+		})
+	}
+	dsMorefs, err := vs.getDatastoreMorefs(ctx, compatibleDatastoreRefs)
+	if err != nil {
+		return nil, err
+	}
+	return dsMorefs, nil
+}
+
+// Verify if the user specified datastore is in the list of compatible datastores.
+func IsUserSpecifiedDatastoreCompatible(dsRefs []mo.Datastore, dsName string) bool {
+	for _, ds := range dsRefs {
+		if ds.Info.GetDatastoreInfo().Name == dsName {
+			return true
+		}
+	}
+	return false
+}
+
+// Get the best fit compatible datastore by free space.
+func GetBestFitCompatibleDatastore(dsRefs []mo.Datastore) string {
+	var curMax int64
+	curMax = -1
+	var index int
+	for i, ds := range dsRefs {
+		dsFreeSpace := ds.Info.GetDatastoreInfo().FreeSpace
+		if dsFreeSpace > curMax {
+			curMax = dsFreeSpace
+			index = i
+		}
+	}
+	return dsRefs[index].Info.GetDatastoreInfo().Name
+}
+
+// Get the datastore morefs.
+func (vs *VSphere) getDatastoreMorefs(ctx context.Context, dsRefs []types.ManagedObjectReference) ([]mo.Datastore, error) {
+	pc := property.DefaultCollector(vs.client.Client)
+	var datastoreMorefs []mo.Datastore
+	err := pc.Retrieve(ctx, dsRefs, []string{DatastoreInfoProperty}, &datastoreMorefs)
+	if err != nil {
+		return nil, err
+	}
+	return datastoreMorefs, nil
+}
+
+// Get all datastores accessible inside the current Kubernetes cluster.
+func (vs *VSphere) getAllAccessibleDatastoresForK8sCluster(ctx context.Context, resourcePool *object.ResourcePool) ([]types.ManagedObjectReference, error) {
```

Comment:
You are assuming that all the datastores within the cluster are accessible to all k8s nodes. This may not be true! You will also have to check if the ESX hosts running the k8s nodes have access to the datastores.